### DISCUSSION TITLE
Multi-stage pipeline

### DISCUSSION THREAD

**Discussion by cveld** [gruntwork-io/knowledge-base#791] [How do I]
With terraform we use a multi-stage pipeline that first establishes a terraform plan, and after review, kicks off terraform apply.
It is fairly easy to transition a terraform plan to a fresh apply stage, by copying the lock file and the plan file, run `init` and then `apply`. 

How can I do the same with terragrunt? 

Additionally, this multi-stage approach should ideally be taking place per group of modules.

E.g.
Group 1
- baseline1
- baseline2

Group 2
- extension1
- extension2

Group 3
- module1
- module2

The stages should be doing the following:
* Plan 1: yield a plan against all groups. validation: no errrors
* Approval check
* Apply 1: apply group 1. validation: no errors
* Plan 2: yield a plan against all groups. validation: no errors, and group 1 should not contain changes
* Approval check
* Apply 2: apply group 2. validation: no errors
* Plan 3: yield a plan against all groups. validation: no errors, and group 1 and 2 should not contain changes.
* Approval check
* Apply 3: apply group 3. validaton: no errors

Is this possible?

---

<ins datetime="2023-12-11T21:30:53Z">
  <p><a href="https://support.gruntwork.io/hc/requests/110638">Tracked in ticket #110638</a></p>
</ins>

**Comment by gitsstewart** [comment]
@cveld Have you already looked through the Terragrunt docs? https://terragrunt.gruntwork.io/docs/ 
Without more information I'm not sure if the auto-init functionality might be the answer to your question: 
https://terragrunt.gruntwork.io/docs/features/auto-init/

**Reply by cveld** [reply]
@gitsstewart I am not sure I can follow your idea here. Are you proposing to disable auto-init? Maybe that is something to try out indeed. With auto-init disabled, I can try to copy the generated files from the plan-stage to the apply-stage and rerun `terragrunt init`. In that way it is parallel to what we are doing for vanilla terraform.
